common/types/json_struct.go

```diff
@@ -112,20 +112,23 @@ func (m *jsonStruct) Equal(other ref.Value) ref.Value {
 	it := m.Iterator()
 	for it.HasNext() == True {
 		key := it.Next()
-		if otherVal := otherMap.Get(key); IsError(otherVal.Type()) {
-			return False
-		} else if thisVal := m.Get(key); IsError(thisVal.Type()) {
+		if otherVal := otherMap.Get(key); IsError(otherVal) {
 			return False
-		} else if thisVal.Equal(otherVal) != True {
+		} else if thisVal := m.Get(key); IsError(thisVal) {
```

Comment:
The `Get()` call will error if the key is not found, but since this is an equality test, absence of the key should yield `false`. The error is propagated if the element comparison yields an error though. Same for lists.

Reply:
Perhaps Get() should return (ref.Value, bool) and be used accordingly?

Reply:
I need to think about it a bit more, but that seems like a reasonable suggestion. I've added a TODO to refine the logic so I'll get back to this in another PR.